Polkadot Protocol > Protocol Components > Accounts > Account Data Structures (#50)
polkadot-developers/polkadot-docs: polkadot-protocol/protocol-components/accounts/account-data-structure.md

---
title: Account Data Structure

> description: TODO

Reviewer: Description is missing

---

# Account Data Structure

Accounts are a foundational element of any blockchain. This page explains how accounts are managed in the Polkadot SDK and how their data structures are utilized to manage the account lifecycle within the runtime.

## Account

The `Account` data type is a generically defined storage map within the [`frame-system`](https://paritytech.github.io/polkadot-sdk/master/src/frame_system/lib.rs.html#900){target=\_blank} pallet:

 --8<-- 'code/polkadot-protocol/protocol-components/accounts/account-id-information.md'

The StorageMap for an Account consists of the following parameters:

> - The first parameter `(_)` is used in macro expansion.
> - `Blake2_128Concat` specifies the hashing algorithm to use.
> - `T::AccountId` is used as the key for over the `AccountInfo<T::Nonce, T::AccountData>` struct.

Reviewer 2: ```suggestion
- The first parameter `(_)` is used in macro expansion
- `Blake2_128Concat` specifies the hashing algorithm to use
- `T::AccountId` is used as the key for over the `AccountInfo<T::Nonce, T::AccountData>` struct
```

!!! note
    See [StorageMap API](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html){target=\_blank} for more details.

## Account Info

The `AccountInfo` for an account is also defined within the [`frame-system`](https://paritytech.github.io/polkadot-sdk/master/src/frame_system/lib.rs.html){target=\_blank} pallet:

--8<-- 'code/polkadot-protocol/protocol-components/accounts/account-info.md'

Each account has an `AccountInfo` structure that includes the following components:

> - `nonce`: Tracks the number of transactions the account has sent.
> - `consumers`: A reference counter indicating the number of other modules that depend on this account's existence.
> - `providers`: A reference counter indicating the number of modules that permit this account to exist.
> - `sufficients`: A reference counter indicating the number of modules that allow this account to exist for their own purposes only.
> - `AccountData`: A configurable structure that can store various types of data specific to the account.

Reviewer 2: ```suggestion
- `nonce` - tracks the number of transactions the account has sent
- `consumers` - a reference counter indicating the number of other modules that depend on this account's existence
- `providers` - a reference counter indicating the number of modules that permit this account to exist
- `sufficients` - a reference counter indicating the number of modules that allow this account to exist for their own purposes only
- `AccountData` - a configurable structure that can store various types of data specific to the account
```

It holds users accountable for the data they store on-chain. If users wish to delete their accounts and reclaim their existential deposit, they must first remove all data stored across all on-chain pallets to decrement the `consumers` counter.

### Sufficients Reference Counter

The `sufficients` reference counter indicates whether an account is self-sufficient and can exist independently. 

For example, in the [`assets`](https://paritytech.github.io/polkadot-sdk/master/pallet_assets/index.html){target=\_blank} pallet, an account may hold a sufficient quantity of specific assets without needing to maintain a native account balance.

### Account Deactivation 

Pallets also include cleanup functions that decrement the `providers` reference counter, marking the account as deactivated within the pallet-managed scope. 

Once both the `providers` and `consumers` reference counters reach zero, the account is considered deactivated by all on-chain pallets.

### Updating Counters

Runtime developers can manage these counters using the:

> -  `inc_consumers()` - Increment the reference counter on an account.

Reviewer: ```suggestion
- `inc_consumers()` - Increment the reference counter on an account.
```

> -  `inc_consumers()` - Increment the reference counter on an account.
> - `dec_consumers()` - Decrement the reference counter on an account.
> - `inc_providers()` - Increment the provider reference counter on an account.
> - `dec_providers()` - Decrement the provider reference counter on an account.
> - `inc_sufficients()` - Increment the self-sufficient reference counter on an account.
> - `dec_sufficients()` - Decrement the sufficients reference counter on an account.

Reviewer 2: ```suggestion
-  `inc_consumers()` - increment the reference counter on an account
- `dec_consumers()` - decrement the reference counter on an account
- `inc_providers()` - increment the provider reference counter on an account
- `dec_providers()` - decrement the provider reference counter on an account
- `inc_sufficients()` - increment the self-sufficient reference counter on an account
- `dec_sufficients()` - decrement the sufficients reference counter on an account
```

!!!note
    All these methods are provided by the [`frame-system`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method){target=\_blank} pallet.

For every increment of a counter during an account's lifecycle, a corresponding decrement should be called to ensure proper management of the account state.

There are also three query functions to ease usage on these counters:

> - [`can_inc_consumer()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_inc_consumer){target=\_blank} to check if an account is ready to be used (`providers` > 0).
> - [`can_dec_provider()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_dec_provider){target=\_blank} to check if an account is no longer referenced in runtime whatsoever (`consumers` == 0) before decrementing providers to 0.
> - [`is_provider_required()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.is_provider_required){target=\_blank} to check if an account has outstanding consumer references (`consumers` > 0).

Reviewer 2: ```suggestion
- [`can_inc_consumer()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_inc_consumer){target=\_blank} to check if an account is ready to be used (`providers` > 0)
- [`can_dec_provider()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_dec_provider){target=\_blank} to check if an account is no longer referenced in runtime whatsoever (`consumers` == 0) before decrementing providers to 0
- [`is_provider_required()`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.is_provider_required){target=\_blank} to check if an account has outstanding consumer references (`consumers` > 0)
```

!!! note
    See [frame-system API](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html){target=\_blank} for more details.

## AccountData Trait and Implementation

The [`AccountInfo`](https://paritytech.github.io/polkadot-sdk/master/frame_system/struct.AccountInfo.html){target=\_blank} can be any structure, provided it meets the requirements of the associated type `AccountData` trait bound defined in the [`frame-system::pallet::Config`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/trait.Config.html){target=\_blank} trait. 

> By default, the Substrate runtime configures `AccountInfo` as defined in the [`pallet-balances`](https://paritytech.github.io/polkadot-sdk/master/pallet_balances/struct.AccountData.html){target=\_blank}.

Reviewer: I've started to just say "runtime" instead "Substrate runtime" for consistency, @eshaben is this fine?

```suggestion
By default, the runtime configures `AccountInfo` as defined in the [`pallet-balances`](https://paritytech.github.io/polkadot-sdk/master/pallet_balances/struct.AccountData.html){target=\_blank}.
```

- **`_`** - used in macro expansion and acts as a placeholder for the storage prefix type. Since the default storage prefix (derived from the pallet and storage item name) is used here, it is represented as `_` (underscore), which tells the macro to insert the default prefix during expansion
- **`Blake2_128Concat`** - the hashing function used for keys in the storage map
- **`T::AccountId`** - the key type for the storage map, representing the account ID, which is used to access the corresponding `AccountInfo<T::Nonce, T::AccountData>` struct in storage
- `AccountInfo<T::Nonce, T::AccountData>` - the value type stored in the map. For each account ID, the map stores an `AccountInfo` struct containing:
    - **`T::Nonce`** - a nonce for the account, which is incremented with each transaction to ensure transaction uniqueness
    - **`T::AccountData`** - custom account data defined by the runtime configuration, which could include balances, locked funds, or other relevant information
- **`ValueQuery`** - a trait that defines how queries to the storage map behave when no value is found, returning the default value if a query for a key returns no result (instead of `None`)

!!! note

>     See the [`StorageMap` rustdocs](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html){target=\_blank} for more details.

Reviewer 3: ```suggestion
    See the [`StorageMap` Rust docs](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html){target=\_blank} for more details.
```

!!!note
    The [System pallet](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method){target=\_blank} provides all these methods.

For every counter increment during an account's lifecycle, a corresponding decrement should be called to ensure proper management of the account state.

There are also three query functions to ease usage on these counters:

- [**`can_inc_consumer()`**](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_inc_consumer){target=\_blank} - to check if an account is ready to be used (`providers` greater than 0)
- [**`can_dec_provider()`**](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.can_dec_provider){target=\_blank} - to check if an account is no longer referenced in the runtime whatsoever (`consumers` equal 0) before decrementing `providers` to 0
- [**`is_provider_required()`**](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.is_provider_required){target=\_blank} to check if an account has outstanding consumer references (`consumers` greater than 0)

!!! note

>     See the [System pallet rustdocs](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html){target=\_blank} for more details.

Reviewer 3: ```suggestion
    See the [System pallet Rust docs](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html){target=\_blank} for more details.
```

---
title: Account Data Structure
description: TODO
---

# Account Data Structure

> !!! warning "Update Notice"

Reviewer: Not sure if we need to keep this in these docs?

Reviewer 2: I think we should remove this

    Please read Substrate to [Polkadot SDK](https://docs.substrate.io/polkadot-sdk/) page first.

> Accounts are a foundational element of any blockchain, and in Substrate, they serve various functions. This article explains how accounts are stored in Substrate and how their data structures are utilized to manage the account lifecycle within the runtime logic.

Reviewer: Someone else can correct me, but I think we are using `Polkadot SDK` and not `Substrate`.

```suggestion
Accounts are a foundational element of any blockchain. This page explains how accounts are managed in the Polkadot SDK and how their data structures are utilized to manage the account lifecycle within the runtime.
```

## Account

> The `Account` data type is a generically defined storage map within the [`frame-system`](https://paritytech.github.io/polkadot-sdk/master/src/frame_system/lib.rs.html) pallet:

Reviewer: IRT we link to the Polkadot SDK, or Rust Docs code?

> ```rs

Reviewer: Not sure it is productive to have this snippet here, in the comment above I linked the struct itself in the codebase, which is more maintainable IMO.

Instead, a simple explanation of how it is a mapping of `AccountId` -> `AccountInfo` would suffice

Reviewer 2: I agree

/// The full account information for a particular account ID.
#[pallet::storage]
#[pallet::getter(fn account)]
pub type Account<T: Config> = StorageMap<
  _,
  Blake2_128Concat,
  T::AccountId,
  AccountInfo<T::Nonce, T::AccountData>,
  ValueQuery,
>;
```

The StorageMap for an Account consists of the following parameters:

- The first parameter `(_)` is used in macro expansion.
- `Blake2_128Concat` specifies the hashing algorithm to use.
- `T::AccountId` is used as the key for over the `AccountInfo<T::Nonce, T::AccountData>` struct.

!!! note
    See [StorageMap API](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html) for more details.

## Account Info

The `AccountInfo` for an account is also defined within the [`frame-system`](https://paritytech.github.io/polkadot-sdk/master/src/frame_system/lib.rs.html) pallet:

> ```rs

Reviewer: If we do choose to leave code snippets, then we should put them in snippets, but I am more leaning towards just linking the line in the code base.

#[derive(Clone, Eq, PartialEq, Default, RuntimeDebug, Encode, Decode)]
pub struct AccountInfo<Nonce, AccountData> {
  /// The number of transactions this account has sent.
  pub nonce: Nonce,
  /// The number of other modules that currently depend on this account's existence. The account
  /// cannot be reaped until this is zero.
  pub consumers: RefCount,
  /// The number of other modules that allow this account to exist. The account may not be reaped
  /// until this and `sufficients` are both zero.
  pub providers: RefCount,
  /// The number of modules that allow this account to exist for their own purposes only. The
  /// account may not be reaped until this and `providers` are both zero.
  pub sufficients: RefCount,
  /// The additional data that belongs to this account. Used to store the balance(s) in a lot of
  /// chains.
  pub data: AccountData,
}
```

> Each account has an `AccountInfo` structure that includes the following components:

Reviewer: I think a list like this is probably better than including the struct, IMHO.

- `AccountData`: A configurable structure that can store various types of data specific to the account.

## Account Reference Counters

Account reference counters track dependencies in the runtime. For instance, if data is stored in a map associated with an account, you wouldn't want to delete the account until the data under its control has been removed.

The `consumers` and `providers` reference counters are intended to work together. For example, in the [`session`](https://docs.rs/pallet-session/latest/pallet_session/) pallet, the consumers counter is incremented when an account sets its session keys before becoming a validator. 

The providers counter must be greater than zero for the consumers counter to be incremented.

### Providers Reference Counters

The `providers` reference counter indicates whether an account is ready to be depended upon. For example, it is incremented when a new account is created with a balance greater than the existential deposit.

The `providers` reference counter prevents Substrate pallets from storing data about an account until the account is active (`providers` > 0)

> ### Consumers Reference Counters

Reviewer: Not sure if necessary here, but maybe we have a bullet list explaining what a provider, consumer, and sufficient is in this context?

Author: Not sure I understand what to add here, could you give me more context?

Reviewer: Ah nvm, seems to be a list above that I missed, all good now!

> The `consumers` reference counter ensures that Substrate pallets don't remove an account until all associated data across pallets is deleted (that is, when `consumers` == 0). 

Reviewer: ```suggestion
The `consumers` reference counter ensures that other runtime pallets don't remove an account until all associated data across pallets is deleted (that is, when `consumers` == 0). 
```

It holds users accountable for the data they store on-chain. If users wish to delete their accounts and reclaim their existential deposit, they must first remove all data stored across all on-chain pallets to decrement the `consumers` counter.

### Sufficients Reference Counter

The `sufficients` reference counter indicates whether an account is self-sufficient and can exist independently. 

For example, in the [`assets`](https://paritytech.github.io/polkadot-sdk/master/pallet_assets/index.html) pallet, an account may hold a sufficient quantity of specific assets without needing to maintain a native account balance.

### Account Deactivation 

Pallets also include cleanup functions that decrement the `providers` reference counter, marking the account as deactivated within the pallet-managed scope. 

Once both the `providers` and `consumers` reference counters reach zero, the account is considered deactivated by all on-chain pallets.

### Updating Counters

> Runtime developers can manage these counters using the `inc_consumers()`, `dec_consumers()`, `inc_providers()`, `dec_providers()`, `inc_sufficients()`, and `dec_sufficients()` methods provided by the `frame-system` pallet. 

Reviewer: I am not sure if we should link the functions themselves, instead link where the functions live, i,e.,: https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method

Reviewer: Although in this case, it is highly unlikely something like `frame_system` would change, at least a bullet list format would be better.

Author: I created a bullet list and added descriptions to the functions listed

For every increment of a counter during an account's lifecycle, a corresponding decrement should be called to ensure proper management of the account state.

There are also three query functions to ease usage on these counters:

> - `can_inc_consumer()` to check if an account is ready to be used (`providers` > 0).

Reviewer: ditto, or link the functions in the docs

Author: Same, I added links to all these three functions from **frame_system::pallet::Pallet**

- `can_dec_provider()` to check if an account is no longer referenced in runtime whatsoever (`consumers` == 0) before decrementing providers to 0.
- `is_provider_required()` to check if an account has outstanding consumer references (`consumers` > 0).

!!! note
    See [frame-system API](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html) for more details.

## AccountData Trait and Implementation

> The `AccountInfo` can be any structure, provided it meets the requirements of the associated type `AccountData` trait bound defined in the [`frame-system::pallet::Config`](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/trait.Config.html) trait. 

Reviewer: Link AccountInfo via Polkadot SDK Docs

Author: I linked AccountInfo to frame_system::AccountInfo , tell me if this is the correct linkl

By default, the Substrate runtime configures `AccountInfo` as defined in the [`pallet-balances`](https://paritytech.github.io/polkadot-sdk/master/pallet_balances/struct.AccountData.html).

> ## Where to Go Next

Author: makes sense, will remove links

- **`ValueQuery`** - a trait that defines how queries to the storage map behave when no value is found, returning the default value if a query for a key returns no result (instead of `None`)

!!! note

>     See the [`StorageMap` rustdocs](https://paritytech.github.io/polkadot-sdk/master/frame_support/storage/types/struct.StorageMap.html){target=\_blank} for more details.

Reviewer 3: I can't leave a suggestion but I think should be Rust docs

- [**`is_provider_required()`**](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html#method.is_provider_required){target=\_blank} to check if an account has outstanding consumer references (`consumers` greater than 0)

!!! note

>     See the [System pallet rustdocs](https://paritytech.github.io/polkadot-sdk/master/frame_system/pallet/struct.Pallet.html){target=\_blank} for more details.

Reviewer 3: Same here - Rust docs
